File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        ArrayList<Integer> list = new ArrayList<>();
        
        for(ListNode ptr = head; ptr != null ; ptr = ptr.next) 
            list.add(ptr.val);
        
        int[] ans = new int[list.size()];
        Stack<Integer> stack = new Stack<>();
        
        for(int i=0;i<ans.length;i++) {
            while(!stack.isEmpty() && list.get(i) > list.get(stack.peek()))
            {
                ans[stack.pop()] = list.get(i);
            }
            stack.push(i);
        }
        
        return ans;
    }
}

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        if not head.next:
            return [0]
        
        stack = []
        ans = []
        index = 0
 
        while head:
            ans.append(0)            
#             poping and writing on ans 
#             stack contains a tuple of index, val
            while stack and stack[-1][1] < head.val:
                curindex, val = stack.pop()
                ans[curindex] = head.val
            
            stack.append((index, head.val))
            index +=1
            head = head.next
        

        
        return ans
                
                
            